### PR TITLE
Initial port of Intro to OpenMP material

### PR DESCRIPTION
**Pull request by steve-crouch**
Ported to our markdown format by @mehtaparabaci. Like with the MPI material, it assumes access to a cluster, although it's also mentioned a local machine installation of an appropriate compiler and OpenMP will work too. It's been linked to the reworked hpc_parallel_intro, with it as a dependency.

### CODE REVIEW

File: high_performance_computing/hpc_openmp/02_intro_openmp.md [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+## OpenMP vs. Low-Level Threading APIs (POSIX Threads)
+
+When it comes to parallel programming with threads, there are two main ways to tackle it: the 
+user-friendly OpenMP and the more intricate Low-Level Threading APIs. In this context, **Low-Level 
+Threading APIs**, refer to interfaces like the Portable Operating System Interface (POSIX), which 
+defines a set of standard functions and interfaces for interacting with operating systems. Each 
+approach has its own advantages and considerations, so let's break them down in simpler terms:
+
+OpenMP serves as a gentle introduction to parallel programming. It offers an easy way to parallelize 
+your code without delving into complex technicalities. This makes it ideal for beginners or anyone
```
Comment (martinjrobins) [inline high_performance_computing/hpc_openmp/02_intro_openmp.md:95]: I don't agree with the point that it is ideal for beginners (in comparison to threads), this implies that as you get more advanced you should swap to using posix threads, which generally isn't the case as you would rarly use raw threads because there are so many good higher-level tools out there for parallel and concurrent programming that you don't normally resort to using threads!

I would just say that openmp is built on threads and provides a higher-level interface that is suitable for parallel programming

File: high_performance_computing/hpc_openmp/02_intro_openmp.md [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+~~~
+Hello World!
+Hello World!
+Hello World!
+Hello World!
+~~~
+
+## OpenMP vs. Low-Level Threading APIs (POSIX Threads)
+
+When it comes to parallel programming with threads, there are two main ways to tackle it: the
```
Comment (martinjrobins) [inline high_performance_computing/hpc_openmp/02_intro_openmp.md:88]: there are many other libraries for parallel programming that are built on threading (e.g. intell TBB)

File: high_performance_computing/hpc_openmp/02_intro_openmp.md [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+- ***User-Friendly:*** OpenMP requires minimal code adjustments, making it a friendly choice for 
+ newcomers.
+- ***Automated Work Distribution:*** It divides tasks among threads, ensuring balanced workloads.
+- ***Wider Applicability:*** OpenMP functions across different systems and programming languages.
+
+For instance, consider a scenario where you have a task that involves doing the same thing over and 
+over again, like processing a bunch of images. With OpenMP, you can split up the work among different 
+threads, so they each handle a part of the task at the same time.
+
+Now, imagine you're a master chef who wants complete control over every ingredient and spice in a
```
Comment (martinjrobins) [inline high_performance_computing/hpc_openmp/02_intro_openmp.md:110]: I think you're comparing apples and oranges here. Might want to consider the advantages of openmp over other parallel programming libraries (e.g. HPX, Intel TBB, C++ standard library algorithms

I would say that posix threads are just a bad choice for parallel programming full stop, they are more suitable for situations where you *arn't* doing parallel programming, i.e. concurrent / async stuff. Theoretically you could get better performance writing parallel stuff using threads, but I don't think it would be worth the large amount of effort (and additional knowledge you would need to gain) and the sacrifice of portability / maintainability.

Comment (martinjrobins) [reply high_performance_computing/hpc_openmp/02_intro_openmp.md:110]: I would say the advantages of openmp are that it is very easy to get started on, easy to use on existing serial code so is good to migrate an existing serial code to parallel, is a defacto standard for C for parallel programming, but that other options exist on different languages (lots of choice for C++, multiprocessing for python, rayon on rust)

File: high_performance_computing/hpc_openmp/04_synchronisation.md [hunk 4]
```diff
@@ -0,0 +1,529 @@
...
+#include <omp.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#define NUM_TIMES 10000
+
+int main(void) {
+    int value = 0;
+
+#pragma omp parallel for
```
Comment (martinjrobins) [inline high_performance_computing/hpc_openmp/04_synchronisation.md:74]: should this pragma be lined up with the for loop?

File: high_performance_computing/hpc_openmp/04_synchronisation.md [hunk 5]
```diff
@@ -0,0 +1,529 @@
...
+
+```c
+#pragma omp parallel
+{
+    int thread_id = omp_get_num_thread();
+
+    /* The initialisation of the look up table is done in parallel */
+    initialise_lookup_table(thread_id);
+
+#pragma omp barrier  /* As all threads depend on the table, we have to wait until all threads
```
Comment (martinjrobins) [inline high_performance_computing/hpc_openmp/04_synchronisation.md:118]: should be tabbed?

File: high_performance_computing/hpc_openmp/04_synchronisation.md [hunk 6]
```diff
@@ -0,0 +1,529 @@
...
+
+```c
+double old_matrix[NX][NY];
+double new_matrix[NX][NY];
+
+#pragma omp parallel for
+for (int i = 0; i < NUM_ITERATIONS; ++i) {
+    int thread_id = omp_get_thread_num();
+    iterate_matrix_solution(old_matrix, new_matrix, thread_id);
+#pragma omp barrier  /* You may want to wait until new_matrix has been updated by all threads */
```
Comment (martinjrobins) [inline high_performance_computing/hpc_openmp/04_synchronisation.md:136]: same here

File: high_performance_computing/hpc_openmp/04_synchronisation.md [hunk 7]
```diff
@@ -0,0 +1,529 @@
...
+The next example builds on the previous example which included a lookup table. In the the modified code, the lookup
+table is written to disk after it has been initialised. This happens in a single region, as only one thread needs to
+write the result to disk.
+
+```c
+#pragma omp parallel
+{
+    int thread_id = omp_get_num_thread();
+    initialise_lookup_table(thread_id);
+#pragma omp barrier  /* Add a barrier to ensure the lookup table is ready to be written to disk */
```
Comment (martinjrobins) [inline high_performance_computing/hpc_openmp/04_synchronisation.md:196]: here too

File: high_performance_computing/hpc_openmp/04_synchronisation.md [hunk 8]
```diff
@@ -0,0 +1,529 @@
...
+```c
+#include <math.h>
+#include <omp.h>
+
+#define NUM_ELEMENTS 10000
+
+int main(int argc, char **argv) {
+  int array[NUM_ELEMENTS] = {0};
+
+#pragma omp parallel for schedule(static)
```
Comment (martinjrobins) [inline high_performance_computing/hpc_openmp/04_synchronisation.md:242]:
```suggestion
  #pragma omp parallel for schedule(static)
```

File: high_performance_computing/hpc_openmp/04_synchronisation.md (hunk 7 above)
Comment (martinjrobins) [reply high_performance_computing/hpc_openmp/04_synchronisation.md:196]:
```suggestion
    #pragma omp barrier  /* Add a barrier to ensure the lookup table is ready to be written to disk */
```

File: high_performance_computing/hpc_openmp/04_synchronisation.md [hunk 9]
```diff
@@ -0,0 +1,529 @@
...
+
+#define NUM_ELEMENTS 1000
+
+int main(int argc, char **argv) {
+  int array[NUM_ELEMENTS] = {0};
+
+  int progress = 0;
+  int output_frequency = NUM_ELEMENTS / 10; /* output every 10% */
+
+#pragma omp parallel for schedule(static)
```
Comment (martinjrobins) [inline high_performance_computing/hpc_openmp/04_synchronisation.md:271]:
```suggestion
  #pragma omp parallel for schedule(static)
```

File: high_performance_computing/hpc_openmp/04_synchronisation.md [hunk 10]
```diff
@@ -0,0 +1,529 @@
...
+  for (int i = 0; i < NUM_ELEMENTS; ++i) {
+    array[i] = log(i) * cos(3.142 * i);
+
+#pragma omp critical
```
Comment (martinjrobins) [inline high_performance_computing/hpc_openmp/04_synchronisation.md:275]:
```suggestion
    #pragma omp critical
```

Comment (martinjrobins) [reply high_performance_computing/hpc_openmp/04_synchronisation.md:275]: more below but I'll stop now

File: high_performance_computing/hpc_openmp/05_hybrid_parallelism.md [hunk 11]
```diff
@@ -0,0 +1,440 @@
...
+
+
+At this point in the lesson, we've introduced the basics you need to get out there and start writing parallel code using
+OpenMP. There is one thing still worth being brought to your attention, and that is *hybrid parallelism*.
+
+:::callout
+## The Message Passing Interface (MPI)
+
+In this episode, we will assume you have some knowledge about the Message Passing Interface (MPI) and that you have a 
+basic understand of how to paralleise code using MPI. If you're not sure, you can think of MPI as being like an OpenMP
```
Comment (martinjrobins) [inline high_performance_computing/hpc_openmp/05_hybrid_parallelism.md:15]: perhaps put the mpi material as a dependency of this section?

**Review by martinjrobins — CHANGES_REQUESTED** [review]
Thanks @steve-crouch! This looks excellent :) My main comments were related to the intro where you compare openmp with threading. I don't think they can really be compared as they are for two different things (parallel vs concurrent programming), and the implication of your comparison was that you would tend to use threading for parallel programming once you got more advanced, which I don't think is the case.

I also think we should be constantly encouraging students to use IDEs (and they do anyway which is good), so need to put some instructions on compiling openmp code in vscode.

Thank you so much to you and Mehtap for putting this together :)

File: high_performance_computing/hpc_openmp/02_intro_openmp.md [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+- **Environment Variables:** These are settings that can be adjusted to influence the behavior of the OpenMP runtime. They provide a way to fine-tune the parallel execution of your program. Setting OpenMP environment variables is typically done similarly to other environment variables for your system. For instance, you can adjust the number of threads to use for a program you are about to execute by specifying the value in the `OMP_NUM_THREADS` environment variable.
+
+
+## Running a Code with OpenMP
+
+Before we delve into specifics of writing code that uses OpenMP, let's first look at how we compile and run an example "Hello World!" OpenMP program that prints this to the console.
+
+ First, log into whichever HPC system you have access to - this could be a group server, or university- or national-level cluster (e.g. Iridis or DiRAC). Alternatively, you may have OpenMP installed on your local machine.
```
[the remark this answers is not in the document]
Comment (steve-crouch) [reply high_performance_computing/hpc_openmp/02_intro_openmp.md]: Agreed! Fixed.

File: high_performance_computing/hpc_openmp/02_intro_openmp.md [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+    {
+        printf("Hello World!\n");
+    }
+}
+~~~
+
+You'll likely want to compile it using a standard compiler such as `gcc`, although this may depend on your system. To enable the creation of multi-threaded code based on OpenMP directives, pass the `-fopenmp` flag to the compiler. This flag indicates that you're compiling an OpenMP program:
+
+~~~bash
+gcc hello_world_omp.c -o hello_world_omp -fopenmp
```
[the remark this answers is not in the document]
Comment (steve-crouch) [reply high_performance_computing/hpc_openmp/02_intro_openmp.md:65]: Good idea, I've added some draft hints in a callout for this at the bottom of the first episode.

File: high_performance_computing/hpc_openmp/02_intro_openmp.md (hunk 1 above)
Comment (steve-crouch) [reply high_performance_computing/hpc_openmp/02_intro_openmp.md]: Yes, I think I agree. The original intent was to compare OpenMP with something that's sufficiently standard but low level, but since POSIX low-level threads aren't used raw anyway this isn't a practical comparison. I've removed this section.

File: high_performance_computing/hpc_openmp/02_intro_openmp.md (hunk 2 above)
Comment (steve-crouch) [reply high_performance_computing/hpc_openmp/02_intro_openmp.md]: Addressed via its removal (but it might be good to eventually have a section discussing alternatives at this level).

File: high_performance_computing/hpc_openmp/02_intro_openmp.md (hunk 3 above)
Comment (steve-crouch) [reply high_performance_computing/hpc_openmp/02_intro_openmp.md]: Addressed as suggested!

File: high_performance_computing/hpc_openmp/04_synchronisation.md (hunk 4 above)
Comment (steve-crouch) [reply high_performance_computing/hpc_openmp/04_synchronisation.md]: Fixed throughout - including those below!

File: high_performance_computing/hpc_openmp/04_synchronisation.md (hunk 5 above)
Comment (steve-crouch) [reply high_performance_computing/hpc_openmp/04_synchronisation.md]: Fixed.

File: high_performance_computing/hpc_openmp/04_synchronisation.md (hunk 6 above)
Comment (steve-crouch) [reply high_performance_computing/hpc_openmp/04_synchronisation.md]: Fixed.

File: high_performance_computing/hpc_openmp/04_synchronisation.md (hunk 7 above)
Comment (steve-crouch) [reply high_performance_computing/hpc_openmp/04_synchronisation.md]: Fixed.

File: high_performance_computing/hpc_openmp/04_synchronisation.md (hunk 9 above)
Comment (steve-crouch) [reply high_performance_computing/hpc_openmp/04_synchronisation.md]: Fixed.

File: high_performance_computing/hpc_openmp/04_synchronisation.md (hunk 10 above)
Comment (steve-crouch) [reply high_performance_computing/hpc_openmp/04_synchronisation.md]: Fixed.

File: high_performance_computing/hpc_openmp/05_hybrid_parallelism.md (hunk 11 above)
Comment (steve-crouch) [reply high_performance_computing/hpc_openmp/05_hybrid_parallelism.md:18]: I've added the MPI material as a dependency for this episode/section, but not sure it's working. At least, it's not showing as a dependency in the material graph that I can see? Perhaps there's a better way to specify it?

Comment (martinjrobins) [reply high_performance_computing/hpc_openmp/05_hybrid_parallelism.md:18]: hmm, maybe this kind of dependecy isn't shown in the graph. I'll merge this in for now and we can look at it later.

**Review by martinjrobins — APPROVED** [review]
all looks good. Thanks @steve-crouch
